Detour/Source/DetourNavMeshQuery.cpp

```diff
@@ -1026,31 +1054,45 @@ dtStatus dtNavMeshQuery::findPath(dtPolyRef startRef, dtPolyRef endRef,
 		*pathCount = 0;
 	
 	// Validate input
-	if (!m_nav->isValidPolyRef(startRef) || !m_nav->isValidPolyRef(endRef) ||
-		!startPos || !endPos || !filter || maxPath <= 0 || !path || !pathCount)
+	if (!m_nav->isValidPolyRef(endRef) ||
+		!startPoses || !endPos || !filter || maxPath <= 0 || !path || !pathCount)
 		return DT_FAILURE | DT_INVALID_PARAM;
 
-	if (startRef == endRef)
+	for (int i = 0; i < numStarts; ++i)
+		if (!m_nav->isValidPolyRef(startRefs[i]))
+			return DT_FAILURE | DT_INVALID_PARAM;
+
+	for (int i = 0; i < numStarts; ++i)
 	{
-		path[0] = startRef;
-		*pathCount = 1;
-		return DT_SUCCESS;
+		if (startRefs[i] == endRef)
+		{
+			path[0] = startRefs[i];
+			*pathCount = 1;
+			if (startIdx)
+				*startIdx = i;
+			return DT_SUCCESS;
+		}
 	}
 	
 	m_nodePool->clear();
 	m_openList->clear();
+	for (int i = 0; i < numStarts; ++i)
+	{
+		// Put each start node at a unique state.  This handles the
+		// case where there are multiple start positions on the same polygon
+		// and lets us easly know which start was used to reach the goal
+		dtNode* startNode = m_nodePool->getNode(startRefs[i], i);
```

Comment:
The number of states is severely limited (it is a 2 bit value). This will not work unfortunately.

Reply:
I didn't notice that, good catch.

Currently, 24 bits are allocated for the pidx, 2 for state, and 3 for flags.  So there are 3 free bits, which would give 32 states and support for 32 goals.

With some additional logic, I could also use the state to just delineate different starts within a polygon instead of putting each start at a unique state.  We could then decide how many of the free bits to allocated to state (if any) based on how many starts per polygon makes sense to support.

Reply:
I don't think it should be necessary to insert all n nodes into the hash table. It should be ok to just insert the first one and only allocate the rest without inserting them in the hash table. I.e.
1. Add a public `dtNodePool::allocNode`
2. Update `dtNodePool::getNode` to use this
3. Allocate nodes directly when inserting start points

Whether all of this complexity is worth it for this feature I am not sure.

With regards to getting the index, it would maybe be possible to use `dtNode::pidx`, to store the index by adding another flag (or similar), again with a complexity trade-off.

Reply:
Gave this a shot.  I also removed the functionality the return which goal was found as there wasn't a clear simple/clean way to do it.

Reply:
I will need to look a little closer at this before I feel comfortable merging it, but I think it looks good.
Apologies for the slowness.

Reply:
I’ve changed this block of code as there aren’t enough bits in the state feild. Did you mean to comment on the PR as a whole?

Reply:
I mean the addition of the `allocNode` and the interactions with potentially not keeping nodes in the hash table.